### PR TITLE
Create clearsky forecast endpoint

### LINKED ISSUE #52: Add route for clear sky output
## Detailed Description

Add a `GET` endpoint `/sites/{site_uuid}/clearsky` to return the clear sky outputs over the duration of the forecasts from `/sites/{site_uuid}/pv_forecast` 

## Context

For the [pv-sites-mobile](https://github.com/openclimatefix/pv-sites-mobile) project, we'd like to be able to compare forecasted outputs with the output of a clear sky. This information, when presented in a graph, will allow our users to easily contextualize the forecast for the upcoming day, giving a visual benchmark for the amount of solar power they'll receive. 

## Possible Implementation

1. add [pvlib](https://pvlib-python.readthedocs.io/en/stable/) as a dependency using `poetry`
2. add a `GET` endpoint `/sites/{site_uuid}/clearsky` 
3. retrieve site details using the `site_uuid` 
4. initialize a [`location`](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.location.Location.html#pvlib-location-location) with pvlib, using relevant site details
5. get clear sky estimates of `ghi, dni, dhi` using the [`get_clearsky`](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.location.Location.get_clearsky.html#pvlib.location.Location.get_clearsky) function
6. calculate total irradiance using the [`get_total_irradiance`](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.irradiance.get_total_irradiance.html) function, using `surface_tilt` and `surface_azimuth` from our site details and passing in `solar_zenith` and `solar_azimuth` from the [`Location.get_solarposition`](https://pvlib-python.readthedocs.io/en/stable/reference/generated/pvlib.location.Location.get_solarposition.html) function
7. utilize total irradiance to calculate clear sky output and return 

Currently, we're not really clear on how to carry out that last step—specifically, calculating output from irradiance. We've looked into calculating output using [`pvlib.pvsystem.pvwatts_dc`](https://pvlib-python.readthedocs.io/en/stable/reference/pv_modeling/generated/pvlib.pvsystem.pvwatts_dc.html#pvlib.pvsystem.pvwatts_dc), but we're not sure about how to retrieve the required parameters associated with cell temperature.

### PR DESCRIPTION
**Pull request by ericcccsliu**
# Pull Request

## Description

Follow-up from #53, closes #52 

Adds an endpoint to get an estimate of the output of a site under a clear sky. 

## How Has This Been Tested?

Manually tested with fake data and ensured that route worked. Looked over the response and did a sanity check to see if the data was reasonable. Added additional tests using pydantic models. Quick plot of data below. 

image

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by peterdudfield** [comment]
Could you do a plot with the actuals and the clear sky index? Hopefully on the same time axis? This would be useful to see if the scales are right.

File: pv_site_api/main.py [hunk 1]
```diff
@@ -111,21 +119,7 @@ def get_sites(
...
-                region=site.region,
-                dno=site.dno,
-                gsp=site.gsp,
-                latitude=site.latitude,
-                longitude=site.longitude,
-                installed_capacity_kw=site.capacity_kw,
-                created_utc=site.created_utc,
-            )
-        )
+        pv_sites.append(site_to_pydantic(site))
```
Comment (peterdudfield) [inline pv_site_api/main.py:122]: how come you have this `site_to_pydantic` function? I think `PVSiteMetadata` is already pydantic object

File: pv_site_api/main.py [hunk 2]
```diff
@@ -291,6 +285,56 @@ def get_pv_forecast_many_sites(
...
+        solar_azimuth=solar_position["azimuth"],
+    )
+
+    # Value for "gamma_pdc" is set to the fixed temp. coeff. used in PVWatts V1
+    # @TODO: allow differing inverter and module capacities
+    # addressed in https://github.com/openclimatefix/pv-site-api/issues/54
+    pv_system = pvsystem.PVSystem(
+        surface_tilt=tilt,
+        surface_azimuth=orientation,
+        module_parameters={"pdc0": site.installed_capacity_kw, "gamma_pdc": -0.005},
```
Comment (ericcccsliu) [inline pv_site_api/main.py:326]: setting `"pdc0"` to `site.installed_capacity_kw` is actually not quite correct here—this should be the nameplate capacity of the solar modules, or the power outputted under irradiance of 1000w/m^2. Unfortunately, it seems like this value is actually completely unknown—is there a suitable constant that we can use as an assumption?

**Comment by ericcccsliu** [comment]
> Could you do a plot with the actuals and the clear sky index? Hopefully on the same time axis? This would be useful to see if the scales are right.

Did a plot using site `725a8670-d012-474d-b901-1179f43e7182`—the actuals were a bit tough to manage due to different intervals between datapoints, so I did a plot against the forecast: 

plot_clearsky

The clearsky estimates actually come out to be less than the forecasted output, which seems counterintuitive to me—ideally, it seems like the clearsky values should be greater. 

One issue in my code is actually mentioned [here](https://github.com/openclimatefix/pv-site-api/pull/55/files#r1145640370), where the current value I'm passing in for `pdc0` for our solar module model isn't quite correct. Having an accurate value would likely fix this issue, but is there some suitable constant that we can use here?

**Comment by peterdudfield** [comment]
> > Could you do a plot with the actuals and the clear sky index? Hopefully on the same time axis? This would be useful to see if the scales are right.
> 
> Did a plot using site `725a8670-d012-474d-b901-1179f43e7182`—the actuals were a bit tough to manage due to different intervals between datapoints, so I did a plot against the forecast:
> 
> plot_clearsky
> 
> The clearsky estimates actually come out to be less than the forecasted output, which seems counterintuitive to me—ideally, it seems like the clearsky values should be greater.
> 
> One issue in my code is actually mentioned [here](https://github.com/openclimatefix/pv-site-api/pull/55/files#r1145640370), where the current value I'm passing in for `pdc0` for our solar module model isn't quite correct. Having an accurate value would likely fix this issue, but is there some suitable constant that we can use here?

Hey, Thanks for doing the plot. Probably worth doing the actuals, because the forecast could be off. 
How come it harder?

yea i agree the clear sky should be higher. Perhaps we could try a constant and see what works. It should be something like 6, i.e 6m^2 of solar panels makes about 1KW. Taking into account effeciency, we probably should take the number down to like ~2. I know thats a bit of an estimate, but perhaps useful to use. 

We did a bit of work estimateing the capacity. @simlmx did this. For this particular site its around ~3 KW. Do you mind plotting that? @simlmx this might be a good reason for use to add those inferred values to the database

**Comment by ericcccsliu** [comment]
> 3 KW. Do you mind plotting that? @simlmx this might be a good reason for use to add those inferred values to the database

I went ahead and changed the nameplate capacity to 6, which makes the clearsky estimate look more reasonable. I also went ahead and graphed actuals, and the capacity of the site, which should be 4 kW. 

image

Taking the nameplate capacity down to 2 didn't quite work—it just brought the clearsky estimates much lower than the forecasts and the actuals. The listed capacity of the site was ~4 kW, so 6 is 1.5x the capacity. I thought that perhaps 1.5x might be a good constant multiplier for our nameplate capacity. I went ahead and graphed the forecasts, actuals, and clearsky estimates for a couple other sites using that multiplier of 1.5x: 

image
image

**Comment by peterdudfield** [comment]
@ericcccsliu discussed with @simlmx about estimateing the capacity, but lets leave that for now. 

Ready to merge, do you mind fixing the file conflicts, and then we can get this live

**Review by peterdudfield — APPROVED** [review]

**Comment by peterdudfield** [comment]
looks good @ericcccsliu let me know when your ready to merge

**Comment by ericcccsliu** [comment]
Ready to merge now! As a final small change, I removed the UTC offset from the datetimes returned by the clearsky endpoint, as the datetimes returned by the forecast and actual endpoints do not include UTC offsets.

**Comment by ericcccsliu** [comment]
@peterdudfield ready to merge now! As a final small change, I removed the UTC offset from the datetimes returned by the clearsky endpoint, as the datetimes returned by the forecast and actual endpoints do not include UTC offsets.

**Pull request merged**